proposals/p2360.md

```diff
@@ -0,0 +1,291 @@
+and `MyType as MyInterface` are equivalent.
+
+### Tuple and struct literals
+
+A tuple literal such as `(1, 2, 3)` produces a tuple value. Its type cannot be
+written directly as a literal: the literal `(i32, i32, i32)` also produces a
+tuple value, not a type value. However, the type of `(1, 2, 3)` is easy to
+express: it is the result of implicitly converting `(i32, i32, i32)` to a type,
+so for example `(i32, i32, i32) as Type` evaluates to the type of `(1, 2, 3)`.
+
+There is no conversion from a tuple type to a tuple value, because tuple types
+are values of type `Type`, and the type `Type` does not implicitly convert to a
+tuple type. For metaprogramming, it is likely that we will benefit from having a
+mechanism to convert a tuple type into a tuple of types, but this is likely to
+be possible using a variadic:
+
+```
+@'''
+Takes a tuple of values. Returns a tuple containing the types of those values.
+'''
+fn TupleTypeToTupleOfTypes[Types:! Type,...](x: (Types,...)) -> auto {
+  return (Types,...);
+}
+```
+
+### Name lookup into values with constrained types
+
+Given a generic function with a generic type parameter, uses of that parameter
+will be implicitly converted to type `Type`. For example:
+
+```
+fn F[T:! Printable](x: T) { x.Print(); }
+```
+
+... is equivalent to ...
+
+```
+fn F[T:! Printable](x: T as Type) { x.Print(); }
+```
+
+As a result, we can no longer describe lookup for `x` as looking into the
+type-of-type, that is, the type of the expression after `x:`, because after
+implicit conversion, that expression is of type `Type`. Instead, lookup will
+look into the type of `x`, which symbolically evaluates to `T`, and will look at
+the value `T` to determine where else to look. Because `T` is symbolically the
+name of a generic type parameter, we will look in the type of `T`, which is
+`Printable`.
+
+## Rationale
+
+-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem)
+    -   Making the behavior of tuples and empty structs more like other types is
+        easier for tooling to handle.
+-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)
+    -   The transition between templates and generics is made smoother by
+        changing compound member access to behave consistently regardless of
+        whether the left-hand operand is a type.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Bottoming out at the same type, `Type`, for all types is likely to be
+        easier to understand. However, having `(i32, i32)` not be the type of a
+        pair of integers, but instead be _implicitly convertible to_ the type of
+        a pair of integers, is likely to be harder to understand. It's unclear
+        which of these will be more dominant, but the new behavior is more
+        consistent.
+
+## Alternatives considered
+
+### Alternative terminology
```

Comment:
You might want to link to the [open discussion on 2022-10-27](https://docs.google.com/document/d/1tEt4iM6vfcY0O0DG0uOEMIbaXcZXlNREc2ChNiEtn_w/edit#heading=h.lecxe8qywdxa) where we discussed these alternatives.

Reply:
Done.